### DISCUSSION TITLE
How can I retrieve a list of subkeys, branch of master key

### DISCUSSION THREAD

**Discussion by ironmask2004** [ra1u/redis-dart#51] [Q&A] [answered]
Hi, thanks for your greate job, its really helpfull and easy to use.

How can I retrieve a list of subkeys  from pattern key (userid:*), then I need to delete  all that sub tree keys from redis server.
for now I'm using the run shell command to get that list, but is there an alternative in your library and how do I use it

var result = await Process.run(
        'redis-cli', ['--scan', '--pattern', 'userId:$userId:*']);
    LineSplitter liness = LineSplitter();
    List<String> keys = liness.convert(result.stdout);

    for (var i = 0; i < keys.length; i++) {
      await _cache.send_object(['EXPIRE', '${keys[i]}', '-1']);
    }

**Comment by ra1u** [comment]
Have you tried with redis command [KEYS](https://redis.io/commands/KEYS)

**Reply by ironmask2004** [reply]
Thanks KEYS  works greate , but they siad its not recommanded  in production system, its mainly for debugging, while SCAN COMMAND is the wright command for retreving a keys by pattern,

**Comment by ra1u** [comment, marked as answer]
Oh. I see! Do you have trouble using scan?

Something like - without compiler

```dart
var iter = "0"

do{
    r = await _cache.send_object(['scan',iter, 'MATCH','userId:$userId:*');
    iter = r[0]; // update new iterator
    var keys = r[1]; 
    // delete keys for received range
    for (var i = 0; i < keys.length; i++) {
        await _cache.send_object(['EXPIRE', '${keys[i]}', '-1']);
    }
} while(iter != "0");
```

**Reply by ironmask2004** [reply]
Thanks, my question has been totally answered,
